File: 1 - metodosBusca/pacoteBusca_JomiHubner/src/exemplos/EstadoMapa.java
package exemplos;

import grafo.Grafo;
import grafo.GrafoNaoDirigido;
import grafo.Vertice;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import busca.Antecessor;
import busca.BuscaLargura;
import busca.Estado;
import busca.Nodo;

/**
 * Representa um estado do mundo para o problema de achar
 * rotas em mapas.
 * 
 * No caso o estado e uma das cidades.
 */
public class EstadoMapa implements Estado, Antecessor {
    
    public String getDescricao() {
        StringBuffer ds = new StringBuffer("Encontra rotas no mapa abaixo (o custo dos caminhos esta entre parenteses):\n");
        for (int i=0; i<nomes.length; i++) {
            ds.append("  saindo de "+nomes[i]+" para ");
            Map<Vertice,Integer> custos = mapa.getVertice(i).getCustoAdjacentes();
            for (Vertice v: custos.keySet()) {
                ds.append(nomes[v.getId()]+"("+custos.get(v)+"), ");
            }
            ds.append("\n");
        }
        return ds.toString();
    }

    /** estado: a cidade corrente */
    int cidade;
    int custo = 1;
    
    public EstadoMapa(int c)  {
        this.cidade = c;
    }
    EstadoMapa(int c, int custo)  {
        this.cidade = c;
        this.custo = custo;
    }
    
    /**
     * verifica se um estado e igual a outro
     */
    public boolean equals(Object o) {
        if (o instanceof EstadoMapa) {
            EstadoMapa e = (EstadoMapa)o;
            return e.cidade == this.cidade;
        }
        return false;
    }
    
    public int hashCode() {
        return new Integer(cidade).hashCode();
    }

    /**
     * verifica se o estado e meta
     */
    public boolean ehMeta() {
        return this.cidade == meta;
    }
    
    /**
     * Custo para geracao de um estado
     */
    public int custo() {
        return custo;
    }
    
    /**
     * gera uma lista de sucessores do nodo.
     */
    public List<Estado> sucessores() {
        List<Estado> suc = new LinkedList<Estado>(); // a lista de sucessores
        Map<Vertice,Integer> custos = mapa.getVertice(cidade).getCustoAdjacentes();
        for (Vertice v: custos.keySet()) {
            suc.add(new EstadoMapa(v.getId(), custos.get(v)));
        }
        return suc;
    }
    
    public List<Estado> antecessores() {
        return sucessores();
    }
    
    public String toString() {
        return nomes[cidade] + " ";
    }
    
    public static int getMeta() {
        return meta;
    }
    public static void setMeta(int cidade) {
        meta = cidade;
    }

    /** informacao estatica (o mapa) */
    public static Grafo mapa = new GrafoNaoDirigido();
    static {
        for (int i=0; i<=15; i++) {
            mapa.criaVertice(i);
        }
        mapa.criaAresta(0,1,3);
        mapa.criaAresta(0,2,6);

        mapa.criaAresta(1,9,3);
        mapa.criaAresta(1,7,3);
        
        mapa.criaAresta(2,13,2);
        mapa.criaAresta(2,14,2);
        mapa.criaAresta(2,3,3);
        mapa.criaAresta(2,6,2);

        mapa.criaAresta(3,4,1);
        mapa.criaAresta(3,5,1);

        mapa.criaAresta(4,5,1);
        mapa.criaAresta(4,8,2);
        mapa.criaAresta(4,11,14); // e -> m

        mapa.criaAresta(5,6,1);
        
        mapa.criaAresta(6,7,2);
        
        mapa.criaAresta(7,8,2);
        mapa.criaAresta(7,9,4);

        mapa.criaAresta(9,10,1);
        mapa.criaAresta(9,12,3);

        mapa.criaAresta(11, 12, 2);
        mapa.criaAresta(11, 15, 1);
    }
    
    // cidades
    // a, b, c, d, e, f, g, h, i, k,  l, m,  n,   o,  p,  x
    // 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15
    public final static char nomes[] = { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'k',  'l', 'm',  'n', 'o',  'p',  'x' };    

    private static int meta;
    
    
    public static void main(String[] a) {
        EstadoMapa inicial = new EstadoMapa(8);
        EstadoMapa.setMeta(15);
        System.out.println(inicial.getDescricao());
        System.out.println("estado inicial= "+inicial);

        Nodo s = new BuscaLargura().busca(inicial);
        if (s != null) {
            System.out.println("solucao = "+s.montaCaminho());
            System.out.println("\toperacoes = "+s.getProfundidade());
            System.out.println("\tcusto = "+s.g());
        }
        
        // test
        /*
        Queue<Nodo> abertos = new PriorityQueue<Nodo>();
        abertos.add(new Nodo(new EstadoMapa(4,50),null));
        abertos.add(new Nodo(new EstadoMapa(1,300),null));
        abertos.add(new Nodo(new EstadoMapa(2,100),null));
        abertos.add(new Nodo(new EstadoMapa(3,200),null));
        abertos.add(new Nodo(new EstadoMapa(5,2),null));
        while (!abertos.isEmpty()) {
            System.out.println(abertos.remove());
        }
        */
       
    }
}

File: 1 - metodosBusca/python/app/exemplos/estado_mapa.py
from typing import List
from app.core.busca import Antecessor, BuscaLargura, Estado
from app.core.grafo import GrafoNaoDirigido


class EstadoMapa(Estado, Antecessor):
    nomes = [
        "a",
        "b",
        "c",
        "d",
        "e",
        "f",
        "g",
        "h",
        "i",
        "k",
        "l",
        "m",
        "n",
        "o",
        "p",
        "x",
    ]
    mapa = GrafoNaoDirigido()
    meta = None

    @staticmethod
    def init_mapa():
        for i in range(16):
            EstadoMapa.mapa.cria_vertice(i)
            EstadoMapa.mapa.cria_aresta(0, 1, 3)
            EstadoMapa.mapa.cria_aresta(0, 2, 6)
            EstadoMapa.mapa.cria_aresta(1, 9, 3)
            EstadoMapa.mapa.cria_aresta(1, 7, 3)
            EstadoMapa.mapa.cria_aresta(2, 13, 2)
            EstadoMapa.mapa.cria_aresta(2, 14, 2)
            EstadoMapa.mapa.cria_aresta(2, 3, 3)
            EstadoMapa.mapa.cria_aresta(2, 6, 2)
            EstadoMapa.mapa.cria_aresta(3, 4, 1)
            EstadoMapa.mapa.cria_aresta(3, 5, 1)
            EstadoMapa.mapa.cria_aresta(4, 5, 1)
            EstadoMapa.mapa.cria_aresta(4, 8, 2)
            EstadoMapa.mapa.cria_aresta(4, 11, 14)
            EstadoMapa.mapa.cria_aresta(5, 6, 1)
            EstadoMapa.mapa.cria_aresta(6, 7, 2)
            EstadoMapa.mapa.cria_aresta(7, 8, 2)
            EstadoMapa.mapa.cria_aresta(7, 9, 4)
            EstadoMapa.mapa.cria_aresta(9, 10, 1)
            EstadoMapa.mapa.cria_aresta(9, 12, 3)
            EstadoMapa.mapa.cria_aresta(11, 12, 2)
            EstadoMapa.mapa.cria_aresta(11, 15, 1)

    def __init__(self, cidade: int, custo: int = 1):
        self.cidade = cidade
        self.custo = custo

    def get_descricao(self) -> str:
        ds = "Encontra rotas no mapa abaixo (o custo dos caminhos esta entre parenteses):\n"
        for i in range(len(self.nomes)):
            ds += f"  saindo de {self.nomes[i]} para "
            custos = self.mapa.get_vertice(i).get_custo_adjacentes()
            for v, custo in custos.items():
                ds += f"{self.nomes[v.get_id()]}({custo}), "
            ds += "\n"
        return ds

    def __eq__(self, other) -> bool:
        if isinstance(other, EstadoMapa):
            return self.cidade == other.cidade
        return False

    def __hash__(self) -> int:
        return hash(self.cidade)

    def eh_meta(self) -> bool:
        return self.cidade == self.meta

    def custo(self) -> int:
        return self.custo

    def sucessores(self) -> List[Estado]:
        suc = []
        custos = self.mapa.get_vertice(self.cidade).get_custo_adjacentes()
        for v, custo in custos.items():
            suc.append(EstadoMapa(v.get_id(), custo))
        return suc

    def antecessores(self) -> List[Estado]:
        return self.sucessores()

    def __str__(self) -> str:
        return f"{self.nomes[self.cidade]} "

    @staticmethod
    def get_meta() -> int:
        return EstadoMapa.meta

    @staticmethod
    def set_meta(cidade: int):
        EstadoMapa.meta = cidade

    @staticmethod
    def main():
        EstadoMapa.init_mapa()
        inicial = EstadoMapa(8)
        EstadoMapa.set_meta(15)
        print(inicial.get_descricao())
        print(f"estado inicial= {inicial}")

        s = BuscaLargura().busca(inicial)
        if s is not None:
            print(f"solucao = {s.monta_caminho()}")
            print(f"\toperacoes = {s.get_profundidade()}")
            print(f"\tcusto = {s.g()}")


if __name__ == "__main__":
    EstadoMapa.init_mapa()
    EstadoMapa.set_meta(15)
    inicial = EstadoMapa(8)
    print(inicial.get_descricao())
    print(f"estado inicial= {inicial}")
    s = BuscaLargura().busca(inicial)
    if s is not None:
        print(f"solucao = {s.monta_caminho()}")
        print(f"\toperacoes = {s.get_profundidade()}")
        print(f"\tcusto = {s.g()}")
